core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java

```diff
@@ -4773,6 +4777,24 @@ private void analyzeCheckConstraint(Table table, QualifiedObjectName name, Scope
             analysis.addCheckConstraints(table, expression);
         }
 
+        private static Expression rewriteToAssignment(Map<Identifier, Expression> assignments, Expression sourceExpression)
```

Comment:
This is still not the right approach. There analyzer should not synthesize expressions that didn't come from the original query.

Instead, we need to:
* During analysis, analyze constraints (types, etc) and record them without modification. We may need to add a scope to represent the frame of reference that will be used to evaluate the expressions.
* During planning, right before the insert/update/etc operation, add a projection that evaluates the constraint for each column that's being written. The constraint should be rewritten using the facilities provided by the planner (`PlanBuilder`, `withScope`, etc).